lib/puppet.rb

```diff
@@ -12,7 +12,6 @@
 # see the bottom of the file for further inclusions
 # Also see the new Vendor support - towards the end
 #
-require 'facter'
```

Comment:
@joshcooper any thoughts on this? Would someone expect that when they `require 'puppet'` to also have `Facter` loaded?

Reply:
I think it's ok to remove, because the default facter impl [requires facter](https://github.com/puppetlabs/puppet/pull/8749/files#diff-a0274c2119ce7505bbfbc688c0fab471caf4a0464e45f6c27cfa3ab92ba55734R11). So we're preserving the behavior of `require 'puppet'` requiring facter.

Reply:
Well, I've decoupled everything that has to do with facter when puppet is required. So with this change, `require 'puppet'` will not load, only when puppet is initialized, facter will also be loaded

```
~/Workspace/puppet  PUP-11217 5s                                                                                                                        
❯ bundle exec irb

cannot load such file -- irbtools/more
2.7.2 :001 >
2.7.2 :002 > require 'puppet'
 => true
2.7.2 :003 > Facter
NameError (uninitialized constant Facter)
2.7.2 :004 > Puppet.initialize_settings([])
 => {}
2.7.2 :005 > Facter
 => Facter
2.7.2 :006 >
```

Reply:
Ah got it. I think it's still ok, because settings will raise if you try to resolve any setting that depends on one of the "root" settings without first calling `initialize_settings`:

```
❯ bx irb
irb(main):001:0> require 'puppet'
=> true
irb(main):002:0> Puppet[:statedir]
Traceback (most recent call last):
       16: from /home/josh/.rbenv/versions/2.7.3/lib/ruby/2.7.0/bundler/cli/exec.rb:28:in `run'
       15: from /home/josh/.rbenv/versions/2.7.3/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `kernel_load'
       14: from /home/josh/.rbenv/versions/2.7.3/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `load'
       13: from /home/josh/.rbenv/versions/2.7.3/bin/irb:23:in `<top (required)>'
       12: from /home/josh/.rbenv/versions/2.7.3/bin/irb:23:in `load'
       11: from /home/josh/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
       10: from (irb):2
        9: from /home/josh/work/puppet/lib/puppet.rb:84:in `[]'
        8: from /home/josh/work/puppet/lib/puppet/settings.rb:172:in `[]'
        7: from /home/josh/work/puppet/lib/puppet/settings.rb:1181:in `value'
        6: from /home/josh/work/puppet/lib/puppet/settings.rb:1209:in `value_sym'
        5: from /home/josh/work/puppet/lib/puppet/settings.rb:1453:in `interpolate'
        4: from /home/josh/work/puppet/lib/puppet/settings.rb:1477:in `lookup_and_convert'
        3: from /home/josh/work/puppet/lib/puppet/settings.rb:1494:in `convert'
        2: from /home/josh/work/puppet/lib/puppet/settings.rb:1494:in `gsub'
        1: from /home/josh/work/puppet/lib/puppet/settings.rb:1505:in `block in convert'
Puppet::Settings::InterpolationError (Error converting value for param 'statedir': Could not find value for $vardir)
```

Another good test is to module rspec tests don't break. Using cron with your PR looks good. rspec-puppet initializes puppet via https://github.com/puppetlabs/rspec-puppet/blob/84a9cfe48257e09061bda565189766cebd03d426/lib/rspec-puppet.rb#L82. That ends up 
calling `Puppet.settings.initialize_app_defaults(app_defaults_for_tests)`, which tries to initialize the log level, `:hook => proc {|value| Puppet::Util::Log.level = value },` and that calls `Puppet.runtime[:facter].debugging(level == :debug)` triggering FacterImpl.new

It'd be great if the runtime services were loaded more explicitly when running tests. Maybe add a `Runtime#load_services` method and call that from `Puppet::Test::TestHelper.before_all_tests`? Not sure if there are load-order issues with that?